### PR TITLE
Multi-Spyre/Stream: Fix default tls_idx for multi-spyre stream initialization

### PR DESCRIPTION
**Pull request by jjhursey**
#### What type of PR is this?

- [x] bug
- [ ] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

* `SpyreGuardImpl::tls_idx` needs to be defaulted to non-zero on local_rank > 0. Otherwise `initializeStreamPoolImpl` will be called multiple times because some threads use the default `0` while others use the `LOCAL_RANK` value (if available).
  * This caused a crash at the non-zero ranks when checking for duplicate stream id registration.
  * By defaulting the `tls_idx` to the `LOCAL_RANK` value, we only call `initializeStreamPoolImpl` once instead of multiple times due to the default `tls_idx=0` and `tls_idx=LOCAL_RANK` which caused the exception to be thrown.

#### Which issue(s) this PR is related to:

Resolves Multi-Spyre support for LLMs. The following Issues/PRs are limited due to this bug.
* https://github.com/torch-spyre/torch-spyre/pull/3490
* https://github.com/torch-spyre/hf-adapters/issues/158

#### Special notes for your reviewer:

This worked for single Spyre runs because `tls_idx` defaulted to `0` which was correct for the "Rank 0" device. This failure only showed up with Multi-Spyre runs at the non-zero ranks. It did not show up in the unit tests that we have, but when running HF Adapters Granite. I was not able to isolate a smaller unit test reproducer, but it has something to do with the `spyre_copy_from` calling `getCurrentStream` with a device id of either `-1` or `0` on the non-zero ranks.

#### Does this PR introduce a user-facing change?

No

#### Additional note:

N/A

### CODE REVIEW

**Review by ani300 — COMMENTED** [review]
Nice root-cause fix — defaulting the thread_local from `LOCAL_RANK` so all threads agree on the device index is the right call, and the `initialized_device_index` guard is a reasonable louder-failure assertion. A few non-blocking suggestions inline; the parser-dedup one is the one I'd most like to see addressed before merge, since a parser divergence could quietly resurrect the exact multi-init split this PR fixes.

File: torch_spyre/csrc/spyre_guard.cpp [hunk 1]
```diff
@@ -113,7 +116,19 @@ c10::DeviceCapability SpyreGuardImpl::getDeviceCapability(
   return cap;
 }
 
-thread_local c10::DeviceIndex SpyreGuardImpl::tls_idx = 0;
+static c10::DeviceIndex default_device_index() {
+  const char* env = std::getenv("LOCAL_RANK");
+  if (env) {
+    char* end;
+    int64_t val = std::strtoll(env, &end, 10);
```
Comment (ani300) [inline torch_spyre/csrc/spyre_guard.cpp]: We now parse `LOCAL_RANK` in two places with two different parsers: here (`strtoll` + end-check) for every thread's thread_local init, and in `module.cpp` `_startRuntime` (`std::atoi`, then it back-fills `tls_idx`). Two parsers for the same env var is exactly the kind of drift that could re-diverge `tls_idx` between the runtime thread and worker threads — i.e. reintroduce the multi-init split this PR is fixing. Could we extract a single `parse_local_rank()` helper (parse + validate + clamp) and call it from both sites?

File: torch_spyre/csrc/spyre_guard.cpp [hunk 2]
```diff
@@ -113,7 +116,19 @@ c10::DeviceCapability SpyreGuardImpl::getDeviceCapability(
...
+    if (end != env) {
+      return static_cast<c10::DeviceIndex>(val);
```
Comment (ani300) [inline torch_spyre/csrc/spyre_guard.cpp]: `c10::DeviceIndex` is `int8_t`, so this cast silently truncates: `LOCAL_RANK=128 -> -128`, `256 -> 0`, and `-1` aliases the "unspecified device" sentinel. `_startRuntime` guards its parse with `logical_device_id < num_devices`, but this path has no bounds/negative check, so an out-of-range rank yields a wrong `tls_idx` instead of a clear error. Physically implausible for Spyre (>=128 local ranks/host), so this is robustness-tier — but worth closing the asymmetry with the checked path, ideally folded into the shared helper above.

File: torch_spyre/csrc/spyre_stream.cpp [hunk 3]
```diff
@@ -311,7 +322,8 @@ void initializeStreamPoolImpl(c10::DeviceIndex device_index) {
         pool.stream_handle_map.find(sid) == pool.stream_handle_map.end(),
         "Host compute stream id ", sid,
         " is already registered; only one Spyre device per process is "
-        "supported");
+        "supported. Device ID ",
```
Comment (ani300) [inline torch_spyre/csrc/spyre_stream.cpp]: Two things here. (1) With the new guard at the top of `initializeStreamPoolImpl` throwing on a second device_index *before* any registration, this pre-existing "already registered" check is now effectively unreachable on that path. (2) The appended value is `pool.initialized_device_index`, which was just assigned `= device_index` at the top — so it prints the *current* index, not the *conflicting original* one, which is misleading if it somehow did fire. Suggest either collapsing to the single new guard, or having the new guard's message report both indices (it has both in scope) and dropping this append.

File: torch_spyre/csrc/spyre_stream.cpp [hunk 4]
```diff
@@ -71,6 +71,9 @@ struct StreamPool {
 
   // Per-device initialization flags
   std::unordered_map<c10::DeviceIndex, std::once_flag> device_init_flags;
+
+  // Tracks the device_index parameter used during initializeStreamPoolImpl
+  c10::DeviceIndex initialized_device_index = -1;
```
Comment (ani300) [inline torch_spyre/csrc/spyre_stream.cpp:79]: Minor: `-1` is overloaded as the "uninitialized" sentinel, but it's also c10's "unspecified device" value. If Impl were ever entered with `device_index == -1`, the assignment `initialized_device_index = device_index` is a no-op and the guard stays disarmed, so a later init with a real index wouldn't trip it. All current entry points remap `-1 -> tls_idx` before reaching Impl, so this is unreachable today — but a plain `bool initialized` flag would be immune to it and reads more clearly.

File: torch_spyre/csrc/spyre_guard.cpp (hunk 1 above)
Comment (jjhursey) [reply torch_spyre/csrc/spyre_guard.cpp]: Addressed in the latest commit. Created a unified `parse_local_rank` function

File: torch_spyre/csrc/spyre_guard.cpp (hunk 2 above)
Comment (jjhursey) [reply torch_spyre/csrc/spyre_guard.cpp]: The new `parse_local_rank` checks the bounds on the value and throws an error

File: torch_spyre/csrc/spyre_stream.cpp (hunk 3 above)
Comment (jjhursey) [reply torch_spyre/csrc/spyre_stream.cpp]: Removed the `Device ID` addition here. I decided to keep the check as a future check for defensive programming, even if we can't reach it today. It should provide some protection if folks are refactoring to know the asserted state here.

File: torch_spyre/csrc/spyre_stream.cpp (hunk 4 above)
Comment (jjhursey) [reply torch_spyre/csrc/spyre_stream.cpp:79]: I kept the `initialized_device_index` for the error message, but added an `initialized` boolean per your suggestion.

**Comment by jjhursey** [comment]
Thanks for the suggestions. I pushed a commit to address them, and verified that the code is working correctly for me

File: torch_spyre/csrc/spyre_guard.cpp [hunk 5]
```diff
@@ -113,7 +152,7 @@ c10::DeviceCapability SpyreGuardImpl::getDeviceCapability(
   return cap;
 }
 
-thread_local c10::DeviceIndex SpyreGuardImpl::tls_idx = 0;
+thread_local c10::DeviceIndex SpyreGuardImpl::tls_idx = parse_local_rank();
```
Comment (JRosenkranz) [inline torch_spyre/csrc/spyre_guard.cpp]: parse_local_rank() throws, but it seeds the non-local thread_local tls_idx here — so a malformed/negative LOCAL_RANK std::terminates the process at the first tls_idx odr-use on a handler-less worker thread instead of surfacing a catchable error; seed via a noexcept wrapper that catches and returns 0, leaving the throwing validation to _startRuntime

Comment (jjhursey) [reply torch_spyre/csrc/spyre_guard.cpp]: Good suggestion. I just pushed a commit for that change.

**Review by thoangtrvn — COMMENTED** [review]
@jjhursey : 
Can we add an integration test to avoid regression? e.g. using torchrun --nproc_per_node=2
- Negative test: LOCAL_RANK=invalid should raise before any tensor operations
- Edge case: LOCAL_RANK explicitly set in Python before import torch_spyre

**Comment by jjhursey** [comment]
@thoangtrvn I added some unit tests along the lines of what you are looking for.

I added them to the existing `spyreccl_backend.py` since it runs with `torchrun` (unlike the other unit tests outside `distributed`). This will allow us to group any future `torchrun` envar checks here.

A tricky part of this check is that Spyre Comms also validates `LOCAL_RANK` and might catch it before `parse_local_rank()`, and it can emit a different exception message string. I built the tests to expect the subprocess to fail with a non-zero exit status. However, checking the stdout/stderr message needs to account for both layers of the stack (torch-spyre and spyre-comms). I made those updates and placed some comments in the test case to reflect this.

**Review by thoangtrvn — APPROVED** [review]
LGTM
